File: src/thermophysicalModels/thermophysicalFunctions/freesteam-2.1/thcond.c
/*
 freesteam - IAPWS-IF97 steam tables library
 Copyright (C) 2004-2009  John Pye
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

/* Appendix B: Recommended Interpolating equation for Industrial Use */
/* see http://www.iapws.org/relguide/thcond.pdf */

#define FREESTEAM_BUILDING_LIB
#include "thcond.h"

#include <math.h>

#define THCOND_TSTAR 647.26
#define THCOND_RHOSTAR 317.7
#define THCOND_KSTAR 1.0

#define THCOND_b0 -0.397070
#define THCOND_b1 0.400302
#define THCOND_b2 1.060000
#define THCOND_B1 -0.171587
#define THCOND_B2 2.392190

#define THCOND_C1 0.642857
#define THCOND_C2 -4.11717
#define THCOND_C3 -6.17937
#define THCOND_C4 0.00308976
#define THCOND_C5 0.0822994
#define THCOND_C6 10.0932

#define THCOND_d1 0.0701309
#define THCOND_d2 0.0118520
#define THCOND_d3 0.00169937
#define THCOND_d4 -1.0200

/* freesteam code */
double freesteam_k_rhoT(double rho, double T){

#define THCOND_a_COUNT 4
	const double THCOND_a[THCOND_a_COUNT] = {
		0.0102811
		,0.0299621
		,0.0156146
		,-0.00422464
	};

	double Tbar = T / THCOND_TSTAR;
	double rhobar = rho / THCOND_RHOSTAR;

	/* fast implementation... minimised calls to 'pow' routine... */

	double Troot = sqrt(Tbar);
	double Tpow = Troot;
	double lam = 0;

	int k;
	for(k = 0; k < THCOND_a_COUNT; ++k) {
		lam += THCOND_a[k] * Tpow;
		Tpow *= Tbar;
	}

	lam += THCOND_b0 + THCOND_b1 * rhobar + THCOND_b2 * exp(THCOND_B1 * SQ(rhobar + THCOND_B2));
	
	double DTbar = fabs(Tbar - 1) + THCOND_C4;
	double DTbarpow = pow(DTbar, 3./5);
	double Q = 2. + THCOND_C5 / DTbarpow;

	double S;
	if(Tbar >= 1){
		S = 1. / DTbar;
	}else{
		S = THCOND_C6 / DTbarpow;
	}

	double rhobar18 = pow(rhobar, 1.8);
	double rhobarQ = pow(rhobar, Q);

	lam += 
		(THCOND_d1 / ipow(Tbar,10) + THCOND_d2) * rhobar18 * 
			exp(THCOND_C1 * (1 - rhobar * rhobar18))
		+ THCOND_d3 * S * rhobarQ *
			exp((Q/(1+Q))*(1 - rhobar*rhobarQ))
		+ THCOND_d4 *
			exp(THCOND_C2 * ipow(Troot,3) + THCOND_C3 / ipow(rhobar,5));

	return THCOND_KSTAR * lam;
}

File: src/thermophysicalModels/thermophysicalFunctions/freesteam-2.1/thcond.py
# -*- coding: utf-8 -*-
# freesteam - IAPWS-IF97 steam tables library
# Copyright (C) 2004-2009  John Pye
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of
# the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the Free
# Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
# Boston, MA 02110-1301, USA.

from math import exp
from freesteam import *

def tc_ptrho(p,T, rho):
	"""
	Revised release on the IAPS Formulation 1985 for the Thermal Conductivity of ordinary water
	IAPWS September 1998
	Page 8

	Converted from Python from the XSteam OpenOffice version.
	"""

	print "rho = %f, T = %f" % (rho, T)

	# ver2.6 Start corrected bug
	if T < 273.15:
		raise RuntimeWarning("T under range")
	elif T < 500 + 273.15:
		if p > 100:
			raise RuntimeWarning("p over range for T<500°C")
	elif T <= 650 + 273.15:
		if p > 70:
			raise RuntimeWarning("p over range for T<650°C")
	elif T <= 800 + 273.15:
		if p > 40:
			raise RuntimeWarning("p over range for T<800°C")
	# ver2.6 End corrected bug

	T = T / 647.26
	rho = rho / 317.7
	lam0 = T ** 0.5 * (0.0102811 + 0.0299621 * T + 0.0156146 * T ** 2 - 0.00422464 * T ** 3)
	lam1 = -0.39707 + 0.400302 * rho + 1.06 * exp(-0.171587 * (rho + 2.39219) ** 2)
	dT = abs(T - 1) + 0.00308976
	Q = 2 + 0.0822994 / dT ** (3. / 5)

	if T >= 1:
		s = 1. / dT
	else:
		s = 10.0932 / dT ** (3. / 5)

	lam2 = (0.0701309 / T ** 10 + 0.011852) * rho ** (9. / 5) * exp(0.642857 * (1 - rho ** (14. / 5))) + 0.00169937 * s * rho ** Q * exp((Q / (1. + Q)) * (1. - rho ** (1. + Q))) - 1.02 * exp(-4.11717 * T ** (3. / 2) - 6.17937 / rho ** 5)

	print "lam0 = %f, lam1 = %f, lam2 = %f" % (lam0, lam1, lam2)

	return lam0 + lam1 + lam2

if __name__=='__main__':
	p = 5
	T = 300
	rho = steam_pT(p * 1e6,T + 273.15).rho
	print "p = %f MPa" % p
	print "T = %f °C" % T
	print "rho = %f kg/m³" % rho
	
	k = tc_ptrho(p, T + 273.15, rho)
	print "k =",k

